src/compiler/tsbuildPublic.ts

```diff
@@ -1459,80 +1506,182 @@ namespace ts {
+                        modifiedTime: buildInfoTime
+                    });
+                }
+                return {
+                    type: UpToDateStatusType.OutputMissing,
+                    missingOutputFileName: buildInfoPath
+                };
+            }
+
+            const buildInfo = Debug.checkDefined(getBuildInfo(state, buildInfoPath, resolvedPath, buildInfoTime));
+            if ((buildInfo.bundle || buildInfo.program) && buildInfo.version !== version) {
+                return {
+                    type: UpToDateStatusType.TsVersionOutputOfDate,
+                    version: buildInfo.version
+                };
+            }
 
-                    // We have an output older than an upstream output - we are out of date
-                    Debug.assert(oldestOutputFileName !== undefined, "Should have an oldest output filename here");
+            if (buildInfo.program) {
+                if ((buildInfo.program as ProgramMultiFileEmitBuildInfo).changeFileSet?.length ||
+                    (!project.options.noEmit && (buildInfo.program as ProgramMultiFileEmitBuildInfo).affectedFilesPendingEmit?.length)) {
                     return {
-                        type: UpToDateStatusType.OutOfDateWithUpstream,
-                        outOfDateOutputFileName: oldestOutputFileName,
-                        newerProjectName: ref.path
+                        type: UpToDateStatusType.OutOfDateBuildInfo,
+                        buildInfoFile: buildInfoPath
                     };
                 }
             }
-        }
 
-        if (missingOutputFileName !== undefined) {
-            return {
-                type: UpToDateStatusType.OutputMissing,
-                missingOutputFileName
-            };
+            oldestOutputFileTime = buildInfoTime;
+            oldestOutputFileName = buildInfoPath;
+            newestDeclarationFileContentChangedTime = buildInfo.program?.dtsChangeTime ? new Date(buildInfo.program.dtsChangeTime) : undefined;
         }
 
-        if (isOutOfDateWithInputs) {
-            return {
-                type: UpToDateStatusType.OutOfDateWithSelf,
-                outOfDateOutputFileName: oldestOutputFileName,
-                newerInputFileName: newestInputFileName
-            };
+        // Check input files
+        let newestInputFileName: string = undefined!;
+        let newestInputFileTime = minimumDate;
+        // Get timestamps of input files
+        for (const inputFile of project.fileNames) {
+            const inputTime = getModifiedTime(state, inputFile);
+            if (inputTime === missingFileModifiedTime) {
+                return {
+                    type: UpToDateStatusType.Unbuildable,
+                    reason: `${inputFile} does not exist`
+                };
+            }
+
+            // If an buildInfo is older than the newest input, we can stop checking
+            if (buildInfoTime && buildInfoTime < inputTime) {

```

Comment:
Checking if buildInfo is older than input file right here to have to iterate through potentially less number of input files